ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/block/BlockManager.java

```diff
@@ -105,9 +110,40 @@ protected SafeFuture<?> doStop() {
   }
 
   @Override
-  public SafeFuture<BlockImportResult> importBlock(final SignedBeaconBlock block) {
+  public SafeFuture<BlockImportResult> importBlock(
+      final SignedBeaconBlock block, final Optional<BroadcastValidation> broadcastValidation) {
     LOG.trace("Preparing to import block: {}", block::toLogString);
-    return doImportBlock(block, Optional.empty());
+
+    // NO broadcast validation, import the old way
+    if (broadcastValidation.isEmpty()) {
+      return doImportBlock(block, Optional.empty(), Optional.empty());
+    }
+
+    final SafeFuture<BlockImportResult> consensusValidationResult = new SafeFuture<>();
+    final SafeFuture<BlockImportResult> importResult =
+        doImportBlock(block, Optional.empty(), Optional.of(consensusValidationResult));
+
+    // we want a future that completes as soon as the consensus validation is done, or intercept any
+    // early import results\exceptions happening before the consensus validation is completed
+    final SafeFuture<BlockImportResult> consensusValidationResultOrImportFailure =
+        consensusValidationResult.or(importResult);
+
+    final SafeFuture<BlockImportResult> finalImportResult =
+        blockBroadcastValidator
+            .validate(block, broadcastValidation.get(), consensusValidationResultOrImportFailure)
+            .thenCompose(
+                broadcastValidationResult ->
+                    switch (broadcastValidationResult) {
+                      case SUCCESS, CONSENSUS_FAILURE -> importResult;
```

Comment:
Could you, please, add a comment why we are ok to continue with `CONSENSUS_FAILURE` here not returning some immediate result.

Reply:
I think you're actually right that there is a but here. Fixed in the upcoming commit.